examples/pytorch/rgcn/link.py

```diff
@@ -1,45 +1,142 @@
-"""
-Differences compared to MichSchli/RelationPrediction
-* Report raw metrics instead of filtered metrics.
-* By default, we use uniform edge sampling instead of neighbor-based edge
-  sampling used in author's code. In practice, we find it achieves similar MRR.
-"""
-
-import argparse
-import torch as th
+import numpy as np
+import torch
 import torch.nn as nn
 import torch.nn.functional as F
-
+import dgl
 from dgl.data.knowledge_graph import FB15k237Dataset
 from dgl.dataloading import GraphDataLoader
+from dgl.nn.pytorch import RelGraphConv
+import argparse
+import tqdm
+
+# for building training/testing graphs
+def get_subset_g(g, mask, num_rels, bidirected=False):
+    src, dst = g.edges()
+    sub_src = src[mask]
+    sub_dst = dst[mask]
+    sub_rel = g.edata['etype'][mask]
+
+    if bidirected:
+        sub_src, sub_dst = torch.cat([sub_src, sub_dst]), torch.cat([sub_dst, sub_src])
+        sub_rel = torch.cat([sub_rel, sub_rel + num_rels])
+
+    sub_g = dgl.graph((sub_src, sub_dst), num_nodes=g.num_nodes())
+    sub_g.edata[dgl.ETYPE] = sub_rel
+    return sub_g
+
+class GlobalUniform:
+    def __init__(self, g, sample_size):
+        self.sample_size = sample_size
+        self.eids = np.arange(g.num_edges())
+
+    def sample(self):
+        return torch.from_numpy(np.random.choice(self.eids, self.sample_size))
+
+class NegativeSampler:
+    def __init__(self, k=10):
+        self.k = k
+
+    def sample(self, pos_samples, num_nodes):
+        batch_size = len(pos_samples)
+        neg_batch_size = batch_size * self.k
+        neg_samples = np.tile(pos_samples, (self.k, 1))
+
+        values = np.random.randint(num_nodes, size=neg_batch_size)
+        choices = np.random.uniform(size=neg_batch_size)
+        subj = choices > 0.5
+        obj = choices <= 0.5
+        neg_samples[subj, 0] = values[subj]
+        neg_samples[obj, 2] = values[obj]
+        samples = np.concatenate((pos_samples, neg_samples))
 
-from link_utils import preprocess, SubgraphIterator, calc_mrr
-from model import RGCN
+        # binary labels indicating positive and negative samples
+        labels = np.zeros(batch_size * (self.k + 1), dtype=np.float32)
+        labels[:batch_size] = 1
+
+        return torch.from_numpy(samples), torch.from_numpy(labels)
+
+class SubgraphIterator:
+    def __init__(self, g, num_rels, pos_sampler, sample_size=30000, num_epochs=6000):
```

Comment:
Since there is a single valid value for `pos_sampler`, you can remove this argument.